src/nf/nf_rnn_layer_submodule.f90

```diff
@@ -0,0 +1,176 @@
+submodule(nf_rnn_layer) nf_rnn_layer_submodule
+
+  use nf_activation, only: activation_function
+  use nf_base_layer, only: base_layer
+  use nf_random, only: random_normal
+
+  implicit none
+
+contains
+
+  elemental module function rnn_layer_cons(output_size, activation) &
+    result(res)
+    integer, intent(in) :: output_size
+    class(activation_function), intent(in) :: activation
+    type(rnn_layer) :: res
+
+    res % output_size = output_size
+    res % activation_name = activation % get_name()
+    allocate( res % activation, source = activation )
+
+  end function rnn_layer_cons
+
+
+  pure module subroutine backward(self, input, gradient)
+    class(rnn_layer), intent(in out) :: self
+    real, intent(in) :: input(:)
+    real, intent(in) :: gradient(:)
+    real :: db(self % output_size)
+    real :: dw(self % input_size, self % output_size)
+
+    db = gradient * self % activation % eval_prime(self % z)
+    dw = matmul(reshape(input, [size(input), 1]), reshape(db, [1, size(db)]))
+    self % gradient = matmul(self % weights, db)
+    self % dw = self % dw + dw
```

Comment:
I recently modified these lines in [nf_dense_layer_submodule.f90](https://github.com/modern-fortran/neural-fortran/blob/118f795dbe2850032cc29623b735a8da37eac661/src/nf/nf_dense_layer_submodule.f90#L32) for better performances. The same logic could be done here IMO.